File: bun_backend/server/lib/package-scripts.ts
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";

/** Lockfile → package manager, checked in order. */
const LOCKFILES: Array<[string, string]> = [
  ["bun.lock", "bun"],
  ["bun.lockb", "bun"],
  ["pnpm-lock.yaml", "pnpm"],
  ["yarn.lock", "yarn"],
  ["package-lock.json", "npm"],
]

export type PackageScript = {
  name: string
  script: string
  command: string
}

export type PackageScripts = {
  has_package_json: boolean
  package_manager: string
  scripts: PackageScript[]
}

export function detectPackageManager(projectPath: string): string {
  for (const [file, manager] of LOCKFILES) {
    if (existsSync(join(projectPath, file))) return manager
  }
  return "npm"
}

/** Read the `scripts` of a project's package.json, if it has one. */
export function readPackageScripts(projectPath: string): PackageScripts {
  const packageManager = detectPackageManager(projectPath)
  const missing: PackageScripts = {
    has_package_json: false,
    package_manager: packageManager,
    scripts: [],
  }

  let raw: string
  try {
    raw = readFileSync(join(projectPath, "package.json"), "utf8")
  } catch {
    return missing
  }

  const found = { ...missing, has_package_json: true }

  let parsed: { scripts?: unknown }
  try {
    parsed = JSON.parse(raw) as { scripts?: unknown }
  } catch {
    return found
  }

  const scripts = parsed.scripts
  if (!scripts || typeof scripts !== "object") return found

  return {
    ...found,
    scripts: Object.entries(scripts)
      .filter((entry): entry is [string, string] => typeof entry[1] === "string")
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([name, script]) => ({
        name,
        script,
        command: `${packageManager} run ${name}`,
      })),
  }
}

File: wails_backend/internal/lib/packagescripts.go
package lib

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"

	"wails_backend/internal/types"
)

// lockfileManagers maps a lockfile to its package manager, checked in order.
var lockfileManagers = []struct {
	file    string
	manager string
}{
	{"bun.lock", "bun"},
	{"bun.lockb", "bun"},
	{"pnpm-lock.yaml", "pnpm"},
	{"yarn.lock", "yarn"},
	{"package-lock.json", "npm"},
}

func DetectPackageManager(projectPath string) string {
	for _, entry := range lockfileManagers {
		if _, err := os.Stat(filepath.Join(projectPath, entry.file)); err == nil {
			return entry.manager
		}
	}
	return "npm"
}

// ReadPackageScripts reads the `scripts` of a project's package.json, if it has one.
func ReadPackageScripts(projectPath string) types.PackageScripts {
	out := types.PackageScripts{
		HasPackageJSON: false,
		PackageManager: DetectPackageManager(projectPath),
		Scripts:        []types.PackageScript{},
	}

	data, err := os.ReadFile(filepath.Join(projectPath, "package.json"))
	if err != nil {
		return out
	}
	out.HasPackageJSON = true

	var pkg struct {
		Scripts map[string]string `json:"scripts"`
	}
	if err := json.Unmarshal(data, &pkg); err != nil {
		return out
	}

	names := make([]string, 0, len(pkg.Scripts))
	for name := range pkg.Scripts {
		names = append(names, name)
	}
	sort.Strings(names)

	for _, name := range names {
		out.Scripts = append(out.Scripts, types.PackageScript{
			Name:    name,
			Script:  pkg.Scripts[name],
			Command: out.PackageManager + " run " + name,
		})
	}
	return out
}
